backends/ubpf/ubpfParser.cpp

```diff
@@ -0,0 +1,361 @@
+        builder->emitIndent();
+        if (selectCase->keyset->is<IR::DefaultExpression>()) {
+            hasDefault = true;
+            builder->append("default: ");
+        } else {
+            builder->append("case ");
+            visit(selectCase->keyset);
+            builder->append(": ");
+        }
+        builder->append("goto ");
+        visit(selectCase->state);
+        builder->endOfStatement(true);
+        return false;
+    }
+
+    bool UBPFStateTranslationVisitor::preorder(const IR::SelectExpression* expression) {
+        hasDefault = false;
+        if (expression->select->components.size() != 1) {
+            // TODO: this does not handle correctly tuples
+            ::error("%1%: only supporting a single argument for select", expression->select);
+            return false;
+        }
+        builder->emitIndent();
+        builder->append("switch (");
+        visit(expression->select);
+        builder->append(") ");
+        builder->blockStart();
+
+        for (auto e : expression->selectCases)
+            visit(e);
+
+        if (!hasDefault) {
+            builder->emitIndent();
+            builder->appendFormat("default: goto %s;", IR::ParserState::reject.c_str());
+            builder->newline();
+        }
+
+        builder->blockEnd(true);
+        return false;
+    }
+
+    bool UBPFStateTranslationVisitor::preorder(const IR::Member* expression) {
+        if (expression->expr->is<IR::PathExpression>()) {
+            auto pe = expression->expr->to<IR::PathExpression>();
+            auto decl = state->parser->program->refMap->getDeclaration(pe->path, true);
+            if (decl == state->parser->packet) {
+                builder->append(expression->member);
+                return false;
+            }
+        }
+
+        visit(expression->expr);
+        builder->append(".");
+        builder->append(expression->member);
+        return false;
+    }
+
+    void
+    UBPFStateTranslationVisitor::compileExtractField(
+            const IR::Expression* expr, cstring field, unsigned alignment, EBPF::EBPFType* type) {
+        unsigned widthToExtract = dynamic_cast<EBPF::IHasWidth*>(type)->widthInBits();
+        auto program = state->parser->program;
+
+        if (widthToExtract <= 64) {
+            unsigned lastBitIndex = widthToExtract + alignment - 1;
+            unsigned lastWordIndex = lastBitIndex / 8;
+            unsigned wordsToRead = lastWordIndex + 1;
+            unsigned loadSize;
+
+            const char* helper = nullptr;
+            if (wordsToRead <= 1) {
+                helper = "load_byte";
+                loadSize = 8;
+            } else if (widthToExtract <= 16)  {
+                helper = "load_half";
+                loadSize = 16;
+            } else if (widthToExtract <= 32) {
+                helper = "load_word";
+                loadSize = 32;
+            } else {
+                if (widthToExtract > 64) BUG("Unexpected width %d", widthToExtract);
```

Comment:
This seems to be impossible since you just checked that widthToExtract <= 64 above.

Reply:
I removed this check. It is useless, indeed.